src/ir/analysis/derive_default.rs

```diff
@@ -0,0 +1,301 @@
+            "We shouldn't try and insert {:?} twice because if it was \
+             already in the set, `constrain` should have exited early.",
+            id
+        );
+
+        ConstrainResult::Changed
+    }
+}
+
+impl<'ctx, 'gen> MonotoneFramework for CannotDeriveDefault<'ctx, 'gen> {
+    type Node = ItemId;
+    type Extra = &'ctx BindgenContext<'gen>;
+    type Output = HashSet<ItemId>;
+
+    fn new(ctx: &'ctx BindgenContext<'gen>) -> CannotDeriveDefault<'ctx, 'gen> {
+        let cannot_derive_default = HashSet::new();
+        let dependencies = generate_dependencies(ctx, Self::consider_edge);
+
+        CannotDeriveDefault {
+            ctx,
+            cannot_derive_default,
+            dependencies,
+        }
+    }
+
+    fn initial_worklist(&self) -> Vec<ItemId> {
+        self.ctx.whitelisted_items().iter().cloned().collect()
+    }
+
+    fn constrain(&mut self, id: ItemId) -> ConstrainResult {
+        trace!("constrain: {:?}", id);
+
+        if self.cannot_derive_default.contains(&id) {
+            trace!("    already know it cannot derive Default");
+            return ConstrainResult::Same;
+        }
+
+        let item = self.ctx.resolve_item(id);
+        let ty = match item.as_type() {
+            Some(ty) => ty,
+            None => {
+                trace!("    not a type; ignoring");
+                return ConstrainResult::Same;
+            }
+        };
+
+        if ty.is_opaque(self.ctx, item) {
+            let layout_can_derive = ty.layout(self.ctx).map_or(true, |l| {
+                l.opaque().can_trivially_derive_default(self.ctx, ())
+            });
+            return if layout_can_derive {
+                trace!("    we can trivially derive Default for the layout");
+                ConstrainResult::Same
+            } else {
+                trace!("    we cannot derive Default for the layout");
+                self.insert(id)
+            };
+        }
+
+        if ty.layout(self.ctx).map_or(false, |l| l.align > RUST_DERIVE_IN_ARRAY_LIMIT) {
+            // We have to be conservative: the struct *could* have enough
+            // padding that we emit an array that is longer than
+            // `RUST_DERIVE_IN_ARRAY_LIMIT`. If we moved padding calculations
+            // into the IR and computed them before this analysis, then we could
+            // be precise rather than conservative here.
+            return self.insert(id);
+        }
+
+        match *ty.kind() {
+            // Handle the simple cases. These can derive debug without further
+            // information.
+
+            TypeKind::Function(..) |
+            TypeKind::Int(..) |
+            TypeKind::Float(..) |
+            TypeKind::Complex(..) => {
+                trace!("    simple type that can always derive Default");
+                ConstrainResult::Same
+            }
+
+            TypeKind::Void |
+            TypeKind::Named |
+            TypeKind::TemplateInstantiation(..) |
```

Comment:
Yeah, as mentioned on IRC, I think a template instantiation should derive default if its definition and arguments derive default.